File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        res = 0
        macs = -1

        for i in range(n-1, -1, -1):
            macs = max(prices[i], macs)
            res = max(res, macs - prices[i])
        
        return res

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int dfs(TreeNode root, int macs) {
        if (root == null) return 0;
        int lmacs = Math.max(macs, root.val);
        int check = (root.val >= macs) ? 1 : 0;
        int left = dfs(root.left,lmacs);
        int right = dfs(root.right, lmacs);
        return left + right + check;
    }
    public int goodNodes(TreeNode root) {
        return dfs(root, -101);
    }
}
